solr/core/src/java/org/apache/solr/managed/types/CacheManagerPool.java

```diff
@@ -0,0 +1,329 @@
+      if (value == null || !(value instanceof Number)) {
+        return;
+      }
+      double poolLimitValue = ((Number)value).doubleValue();
+      if (poolLimitValue <= 0) {
+        return;
+      }
+      Function<Map<String, Object>, Double> func = controlledToMonitored.get(poolLimitName);
+      if (func == null) {
+        return;
+      }
+      Double totalValue = func.apply(totalValues);
+      if (totalValue.doubleValue() <= 0.0) {
+        return;
+      }
+      double totalDelta = poolLimitValue - totalValue.doubleValue();
+
+      // dead band to avoid thrashing
+      if (Math.abs(totalDelta / poolLimitValue) < deadBand) {
+        return;
+      }
+
+      List<SolrCache> adjustableComponents = new ArrayList<>();
+      components.forEach((name, component) -> {
+        Map<String, Object> resourceLimits = getResourceLimits((SolrCache) component);
+        Object limit = resourceLimits.get(poolLimitName);
+        // XXX we could attempt here to control eg. ramBytesUsed by adjusting maxSize limit
+        // XXX and vice versa if the current limit is undefined or unsupported
+        if (limit == null || !(limit instanceof Number)) {
+          return;
+        }
+        double currentResourceLimit = ((Number)limit).doubleValue();
+        if (currentResourceLimit <= 0) { // undefined or unsupported
+          return;
+        }
+        adjustableComponents.add(component);
+      });
+      optimize(adjustableComponents, currentValues, poolLimitName, poolLimitValue, totalValue.doubleValue());
+    });
+  }
+
+  /**
+   * Manage all eligible components that support this pool limit.
+   */
+  private void optimize(List<SolrCache> components, Map<String, Map<String, Object>> currentValues, String limitName,
+                        double poolLimitValue, double totalValue) {
+    // changeRatio > 1.0 means there are available free resources
+    // changeRatio < 1.0 means there's shortage of resources
+    final AtomicReference<Double> changeRatio = new AtomicReference<>(poolLimitValue / totalValue);
+
+    // ========================== OPTIMIZATION ==============================
+    // if the situation is not critical (ie. total consumption is less than max)
+    // try to proactively optimize by reducing the size of caches with too high hitRatio
+    // (because a lower hit ratio is still acceptable if it means saving resources) and
+    // expand the size of caches with too low hitRatio
+    final AtomicReference<Double> newTotalValue = new AtomicReference<>(totalValue);
+    components.forEach(component -> {
+      long currentLookups = ((Number)currentValues.get(component.getManagedComponentId().toString()).get(SolrCache.LOOKUPS_PARAM)).longValue();
+      long lastLookups = lookups.computeIfAbsent(component.getManagedComponentId().toString(), k -> 0L);
+      if (currentLookups < lastLookups + lookupDelta) {
+        // too little data, skip the optimization
+        return;
+      }
```

Comment:
What if a commit happened between the last adjustment and now and therefore the current value of `lookups` has been reset?

Reply:
Perhaps we can use cumulative lookups to check delta calculations but use lookups to make sure a minimum number of lookups have actually happened to give us meaningful hit ratio data

Reply:
If the lookups has been reset due to commit we should simply skip any optimization - because we really don't know how useful is the content of the cache at the moment.

Using cumulative stats is a double-edged sword, specifically because they are never reset... Take for example cumulative hit ratio - it's calculated using cumulative_lookups and cumulative_hits. This means that short-term fluctuations are averaged, but it also means that short-term effects of any adjustments are also averaged (ie. they don't seem to affect the hit ratio in short term), which during thee next round of optimization will cause more and more adjustments, and the controller will eventually grossly overshoot its target. This issue becomes more and more serious as the absolute numbers increase because it's more and more difficult to sway this average.